lib/src/index.rs

```diff
@@ -967,64 +967,77 @@ impl PartialOrd for IndexEntryByGeneration<'_> {
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd)]
 struct RevWalkWorkItem<'a> {
     entry: IndexEntryByPosition<'a>,
-    wanted: bool,
+    state: RevWalkWorkItemState,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum RevWalkWorkItemState {
+    // Order matters: Unwanted should appear earlier in the max-heap.
+    Wanted,
+    Unwanted,
+}
+
+impl RevWalkWorkItem<'_> {
+    fn is_wanted(&self) -> bool {
+        self.state == RevWalkWorkItemState::Wanted
+    }
 }
 
 #[derive(Clone)]
 pub struct RevWalk<'a> {
     index: CompositeIndex<'a>,
     items: BinaryHeap<RevWalkWorkItem<'a>>,
-    wanted_boundary_set: HashSet<IndexPosition>,
-    unwanted_boundary_set: HashSet<IndexPosition>,
 }
 
 impl<'a> RevWalk<'a> {
     fn new(index: CompositeIndex<'a>) -> Self {
         Self {
             index,
             items: BinaryHeap::new(),
-            wanted_boundary_set: HashSet::new(),
-            unwanted_boundary_set: HashSet::new(),
         }
     }
 
     fn add_wanted(&mut self, pos: IndexPosition) {
-        if !self.wanted_boundary_set.insert(pos) {
-            return;
-        }
         self.items.push(RevWalkWorkItem {
             entry: IndexEntryByPosition(self.index.entry_by_pos(pos)),
-            wanted: true,
+            state: RevWalkWorkItemState::Wanted,
         });
     }
 
     fn add_unwanted(&mut self, pos: IndexPosition) {
-        if !self.unwanted_boundary_set.insert(pos) {
-            return;
-        }
         self.items.push(RevWalkWorkItem {
             entry: IndexEntryByPosition(self.index.entry_by_pos(pos)),
-            wanted: false,
+            state: RevWalkWorkItemState::Unwanted,
         });
     }
+
+    fn pop_eq(&mut self, entry: &IndexEntry<'_>) -> Option<RevWalkWorkItem<'a>> {
+        if let Some(x) = self.items.peek() {
+            (&x.entry.0 == entry).then(|| self.items.pop().unwrap())
+        } else {
+            None
+        }
+    }
+
+    fn skip_while_eq(&mut self, entry: &IndexEntry<'_>) {
+        while self.pop_eq(entry).is_some() {
+            continue;
+        }
+    }
 }
 
 impl<'a> Iterator for RevWalk<'a> {
     type Item = IndexEntry<'a>;
 
     fn next(&mut self) -> Option<Self::Item> {
         while let Some(item) = self.items.pop() {
-            if item.wanted {
-                self.wanted_boundary_set.remove(&item.entry.0.pos);
-                if self.unwanted_boundary_set.contains(&item.entry.0.pos) {
-                    continue;
-                }
+            self.skip_while_eq(&item.entry.0);
```

Comment:
Are there cases where you would want to pop an item without also skipping equal elements? If not, it seems better to extract a helper for popping the next element and discarding duplicates.

Reply:
Ah, yes. I should've mentioned that in the commit message.
A RevWalk that supports generation range filtering will propagate every `Wanted(gen)` as `Wanted(gen + 1)`:
```rust
                while let Some(x) = self.queue.pop_eq(&item.entry.0) {
                    if let RevWalkWorkItemState::Wanted(gen) = x.state {
                        if known_gen != gen {
                            self.queue.push_wanted_parents(&item.entry.0, gen + 1);
                            known_gen = gen;
```

Reply:
Thanks for explaining. Sounds good to have it this way if it's planning for the future.